backend/compact-connect/docs/design/README.md

```diff
@@ -31,6 +57,82 @@ efficient processing. A lambda receives messages from the SQS queue. Each messag
+permissions will be represented by scopes in an access token. See
+[Implementation of permissions](#implementation-of-permissions) for a detailed explanation of the design for exactly
+how permissions are stored and translated into scopes.
+
+#### Compact Executive Directors and Staff
+
+Compact ED level staff can have permission to read all compact data as well as to create and manage users and their
+permissions. They can grant other users the ability to write data for a particular jurisdiction and to create more
+users associated with a particular jurisdiction. They can also delete any user within their compact, so long as that
+user does not have permissions associated with a different compact.
+
+#### Board Executive Directors and Staff
+
+Board ED level staff can have permission to read all compact data, write data to for their own jurisdiction, and to
+create more users that have permissions within their own jurisdiction. They can also delete any user within their
+jurisdiction, so long as that user does not have permissions associated with a different compact or jurisdiction.
+
+#### Implementation of Scopes
+
+AWS Cognito integrates with API Gateway to provide
+[authorizers](https://docs.aws.amazon.com/apigateway/latest/developerguide/apigateway-integrate-with-cognito.html) on an
+API that can verify the tokens issued by a given User Pool and to protect access based on scopes belonging to
+[Resource Servers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-define-resource-servers.html)
+associated with that User Pool. In the Staff Users user pool, we represent each compact as its own Resource Server, with
+associated scopes. Unfortunately, because resource servers support only up to 100 scopes each, and we would like to
+control permission to write to or administrate each of more than 50 jurisdictions independently, the combinations would
+require more than 100 scopes per resource server.
+
+To design around the 100 scope limit, we will have to split authorization into two layers: coarse- and fine-grained.
+We can rely on the Cognito authorizers to protect our API endpoints based on fewer coarse-grained scopes, then
+protect the more fine-grained access within the API endpoint logic. The Staff User pool resource servers will are
+configured with `read`, `write`, and `admin` scopes. `read` scopes indicate that the user is allowed to read the
+entire compact's licensee data. `write` and `admin` scopes, however, indicate only that the user is allowed to write
+or administrate _something_ in the compact respectively, thus giving them access to the write or administrative
+API endpoints. We will then rely on the API endpoint logic to refine their access based on the more fine-grained
+access scopes.
+
+To compliment each of the `write` and `admin` scopes, there will be at least one, more specific, scope, to indicate
```

Comment:
Should the [diagram](https://github.com/csg-org/CompactConnect/blob/7f02302977dae3c7374e49ee0ea3cf07b1b047e9/backend/compact-connect/docs/design/users-arch-diagram.pdf) have a line under `aslp` for the generic `write` scope?

Reply:
Nope. Note that the data structure in the drawing is illustrating the mapping of the data (which more closely reflects the actual business logic) to the scopes (which are really more of an implementation detail). There is no real generic write permissison, so there is no generic write field in the db record. Instead, the `aslp/write` scope is derived together with its fine-grained companion, `aslp/al.write` from the actual permission node in the data structure (such as `permissions.aslp.jurisdictions.al.write`)

Reply:
Right, we discussed this. Bear with me for a second, though...(and thank you in advance)...

Am I right to assume that there are endpoints that are designated as write, read, or admin?

For read and admin, isn't it the case that we let the coarse-grained scope authorize access to those read and admin endpoints respectively, even though access is ultimately determined by the granular scopes?

If that's the case, and assuming there _are_ endpoints that are designated as "write", why would we not follow the same model with write?

What makes "write" different?

Reply:
> Am I right to assume that there are endpoints that are designated as write, read, or admin?

Yes.

> What makes "write" different?

'write' and 'admin' are the same in their approach to authorization, each with a coarse-grained scope that controls endpoint access and a fine-grained scope that ultimately determines real access. The only difference between 'write' and 'admin' that I can think of is that 'admin' has fine-grained scopes for jurisdiction-admin and one for compact-admin, whereas 'write' only has fine-grained scopes for jurisdiction-write. There is no logical write permission equivalent to compact-admin (there is no compact-level data to write).

Reply:
Ok, I think I'm 80% there but will table this for a future discussion :)

I think part of my confusion is that I assumed the `read` and `write` compact-level actions are what directly indicated the _coarse-grained_ scopes, but I'm realizing that that's not the case. The compact-level actions exist precisely because there is data that is specific to a _compact_ but _not_ to a jurisdiction that could have "read" or "write" permissions. I've been over-conflating the diagram's json schema with the scopes themselves.